Pure strategy nash equilibrium

cc7768 opened the pull request:
This adds just a simple brute force algorithm for finding _pure_ strategy Nash equilibrium -- Literally just iterates through all possible action profiles and stores them if they satisfy `is_nash`.

Has a very simple 2 player prisoner's dilemma test.

oyamad commented:
@cc7768 Thanks!

It seems the type of `ne` fails to be inferred. The following seems to work:

```jl
function pure_strategy_NE{N}(nfg::NormalFormGame{N})
    ...
    ne = Array(NTuple{N,Int}, 0)
    ...
end
```

Or we could add a type alias `PureActionProfile{N}`
```jl
typealias PureActionProfile{N} NTuple{N,PureAction}
```

oyamad commented:
Should

```jl
typealias PureAction Integer
```

in [normal_form_game.jl](https://github.com/QuantEcon/Games.jl/blob/589df643614e67cfa725b30daad255e2b23d5b90/src/normal_form_game.jl#L10) be

```jl
typealias PureAction Int
```

?

spencerlyon2 commented:
@oyamad if we want to use `PureAction` like it is a concrete type then aliasing it to `Integer` will not work. 

That being said, I looked through normal_form_game.jl and it seems to me that we only ever use `PureAction` as the type of a function argument, not as a type parameter or to construct arrays. I like that right now it allows for any Integer type and think we should continue to support this. If we do, we just need to be careful that internal uses of `PureAction` won't result in abstract type parameters.

spencerlyon2 commented on src/nash_equilibrium.jl:
```diff
@@ -0,0 +1,23 @@
+    # For each action profile check whether it is NE
+    as = CartesianRange(na)
+    for a in as
+        _a = a.I::NTuple{np, Int}
```
Having to insert type checks is never a great thing because (I believe) it incurs minor runtime cost.

It might be better to add a method `is_nash{N}(::NormalFormGame{N}, ::CartesianIndex{N})` and then we can skip the `_a` bit all together

oyamad commented:
@spencerlyon2 I see, thanks.

oyamad commented:
Some thoughts on the function name.

0. I would remove "strategy". This terminology is reserved for strategies in other classes of games, such as extensive form games. Just `pure_NE` or `pure_nash` should be fine.

0. There are other methods that aim to find pure NEs, so I think it is better to include "brute force" in the function name, as `pure_NE_brute` or `pure_nash_brute`.

oyamad commented:
Maybe better to create a new file `test_nash_equilibrium.jl`?

cc7768 commented:
Excellent. Thanks for the feedback.

1. I'll create a new test file and put the test in there -- Eventually we should probably test slightly more complicated games (like maybe N player Cournot).
2. I will rename to remove the word "strategy" from the function name. Once we have better methods then we can change the name of this function to `pure_NE_brute` (or remove it completely), but for now since it is the only option we might as well keep it as `pure_NE`. Though I'm happy to rethink this if people disagree.
3. I'll find a way to get rid of that type check.
4. I didn't add the type alias, but it sounds like you two agree that it is ok now?

oyamad commented:
I think I would prefer `pure_nash` to `pure_NE`, consistent with the method `is_nash`. Or we could change `is_nash` to `is_NE`, keeping this `pure_NE`.

cc7768 commented:
With respect to what @spencerlyon2 said earlier about not using `PureAction` as a concrete type -- It seems like I am actually doing this in this piece of code. Note in `pure_nash` I use `ne = Array(PureActionProfile, 0)` where `PureActionProfile` is a type alias for `NTuple{N, PureAction}` and `PureAction` is a type alias for `Integer`. Thoughts on best way to address this? I like the aliases that @oyamad has created -- It makes writing the code very intuitive.

Also, removing the type check didn't change anything (as in we don't need it). The type of `::CartesianIndex.I` is `NTuple{...}`, so the check could be removed without a problem.

Moved the test to `test_pure_nash`

Changed function name to `pure_nash`. I agree that the consistency is worthwhile.

oyamad commented:
The following seems to work:

```jl
typealias PureAction Integer
typealias PureActionProfile{N,T<:PureAction} NTuple{N,T}

function pure_nash{N}(nfg::NormalFormGame{N})
    ...
    ne = Array(PureActionProfile{N,Int}, 0)
    ...
end
```

If we follow this approach, should

```jl
typealias PureAction Integer
```

be rather

```jl
typealias AbstractPureAction Integer
```

?

In that case, how could we consistently define `*` and `Abstract*` type aliases?

oyamad commented on src/nash_equilibrium.jl:
```diff
@@ -0,0 +1,26 @@
+"""
+Finds all pure strategy Nash equilibrium for a normal form
```
strategy -> action
equilibrium -> equilibria
(or just "pure Nash equilibria")

oyamad commented on src/nash_equilibrium.jl:
```diff
@@ -0,0 +1,26 @@
+"""
+Finds all pure strategy Nash equilibrium for a normal form
+game. It returns an empty array if there are no nash
```
are -> is
no nash -> no pure Nash or no pure action Nash

cc7768 commented:
Resolved documentation wording.

I like the solution you suggest for the type problem. I would be happy with not defining the `Abstract*` alias at all (and just using `PureAction`). It will mostly be used for function arguments (not array allocation). We could write a small note in dev docs that makes note of this -- "If you need to allocate array of `PureActionProfile` then make sure to specify its type arguments to avoid the issues discussed above." Thoughts?

oyamad commented:
Fine with me.

cc7768 commented:
@oyamad Is this ready to merge -- The only thing undone from our discussion is a note in the dev docs (but there is a note under https://github.com/QuantEcon/Games.jl/issues/17)

oyamad commented:
Maybe rename `nash_equilibrium.jl` to `pure_nash.jl`?

Otherwise I think this is ready to merge.

cc7768 commented:
I renamed the file. I am ready for this to be merged to master. I will wait for you to merge it in order to make sure that we agree.

Thanks for the feedback on this PR @oyamad and @spencerlyon2! I think it is much better than the original version.

oyamad commented:
@cc7768 Thanks, please do merge.
